### ISSUE TITLE
routeChangeComplete event called on page load for routes with params

### ISSUE THREAD

**Issue by anthonyshort** [vercel/next.js#11639]
## Describe the bug

When ASO is enabled and you navigate to pages with params the `routeChangeComplete` is called. On page load:

- `router.query` is `{}` on first load, but [apparently that's expected](https://nextjs.org/docs/api-reference/next/router#router-object).
- `routeChangeComplete` is called.

This is a problem because `routeChangeComplete` isn't called when loading pages that don't have parameters. This inconsistency makes things difficult. 

In my case I'm trying to call analytics events on route change and page load. The page view event is called on page load (using an effect) but then the `routeChangeComplete` is immediately called which fires the page view again. 

## To Reproduce

[Reproduction example here](https://github.com/anthonyshort/next-bug-example)

Steps to reproduce:

1. Start the server
2. Go to `/` and see that `route change` is NOT logged
3. Add any query param, e.g. `/?id=1`. Reload the page and notice `route change` is called
4. Go to `/posts/1` and see that `route change` is logged on page load
5. Add getInitialProps to _app and this doesn't happen any more

## Expected behavior

I would expect either:

1. Route change is always called on page load while ASO is enabled.
2. Route change is not called on page load for pages with params

(I'd also not expect the query object to be empty, but that's probably a separate issue)

## Screenshots

N/A

## System information

- OS: macOS
- Browser (if applies) Chrome
- Version of Next.js: 9.3.4

## Additional context

N/A

**Comment by anthonyshort** [comment]
@timneutkens I can probably take a look into it myself if I can get pointed in the right direction.

**Comment by Yankovsky** [comment]
Same problem here.
routeChangeComplete event is called because this evaluates to true https://github.com/vercel/next.js/blob/a0c683208abdce84934061f6efd2b90de52d99b9/packages/next/client/index.js#L103
Additional condition was probably firstly introduced here https://github.com/vercel/next.js/commit/45832e44219b281eab5569374f3b2cb2f651e755
I don't know what `hydrateProps.__N_SSG` means, probably **Not Static Site Generation**?

**Comment by Yankovsky** [comment]
I've tried local development and static export and in both cases __N_SSG was true.

**Comment by Yankovsky** [comment]
Similar problem https://github.com/vercel/next.js/issues/15082

**Comment by Yankovsky** [comment]
Maybe someone from core team can explain what was intended here.

**Comment by valse** [comment]
Hi, check it out this discussion #12306 about avoid double analytics views on page load

**Comment by valse** [comment]
However this is a big problem when you must trigger something on change routing only: I have this need and at the moment I can't find a workaround to skip the first "bad" `routeChangeComplete` trigger :(

**Comment by Yankovsky** [comment]
@valse that's great, but it is a workaround. It would be great to understand reasons behind the code and to fix root problem.

**Comment by ypresto** [comment]
Still exists in next v10.2.3

**Comment by ArekBartnik** [comment]
Still present in the latest version - v11.1.0.

@leerob would you help? This is an issue only someone from the core team can help resolve.
It's a pretty nasty problem for sites using analytics.

**Comment by GiancarlosIO** [comment]
😢

**Comment by barberdt** [comment]
Adding my recent experience here as well. After seeing wildly inaccurate "page view" events in our even tracking system I determined it was because of this same issue. Completely agnostic of query params, `routeChangeComplete` was firing on the initial page load for all of our SSG pages but not for our SSR pages. I too have now determined the above `router.isReady` to be the best workaround.

To echo what's already been said here, the primary issue here is that, based on code and documentation, I still can't tell what's "expected" here vs. what may be a bug in the router. I guess my expectation would be that `routeChange<anything>` wouldn't fire at all on initial page load for any type of page, but rather only once user interaction or my own code has triggered a history change.

**Comment by and-poulsen** [comment]
Is there any news on this?

**Comment by mgr34** [comment]
Is there any news on this issue? And could I not just use `useRef` and update the `ref.current` property each time, and only track if `!==`?

**Comment by tpreusse** [comment]
@yousefcisco watchout Next v12.2 changed the behaviour again: https://github.com/vercel/next.js/pull/37593 — in our case with middleware and incremental ssg pages we no longer got `routeChangeComplete` events for the initial page view.

New solution:
```js
let lastTrackedPath
const track = path => {
  // avoid double tracking
  // - in case next router becomes ready and triggers a routeChangeComplete
  if (path === lastTrackedPath) {
    return
  }
  lastTrackedPath = path
  // call to analytics api
}

const Track = () => {
  useEffect(() => {
    if (router.isReady) {
      track(router.asPath)
    }
  }, [router.isReady])
  useEffect(() => {
    const onRouteChangeComplete = (asPath) => {
      track(asPath)
    }
    router.events.on('routeChangeComplete', onRouteChangeComplete)
    return () => {
      router.events.off('routeChangeComplete', onRouteChangeComplete)
    }
  }, [])
  return null
}
```

@shuding maybe it's a bug that `routeChangeComplete` is no longer triggered since `v12.2` for SSG pages in some circumstances. To be safe and more feature proof we will now listen to `isReady` changes and `routeChangeComplete` and deduplicate with a custom check.

**Comment by deckgl5** [comment]
@tpreusse https://openjs.space/questions/how-to-update-router-query-in-next-js-without-triggering-a-page-change-event

**Comment by marcoshernanz** [comment]
Thanks for reporting this. This was fixed in Next.js 12.2.0 by #37815, which prevents hydration-only query updates from emitting route change events. I verified the original reproduction on current stable and canary: direct loads of `/?id=1` and `/posts/1` no longer emit `routeChangeComplete`, while client-side navigation still does. Closing as fixed.
